### DISCUSSION TITLE
Using Custom Datasets and Models

### DISCUSSION THREAD

**Discussion by ai21mtech13002** [yoshitomo-matsubara/torchdistill#254] [Q&A] [answered]
Hi Yoshitomo,

This is an excellent tool.  I was planning to use this toolkit on the VHR-10 dataset and a pretrained Faster RCNN (trained using MMDetection). Can you please give a sample on how to perfrom KD using a custom dataset and a custom model for object detection?

PS: I tried to look into the demo notebooks, but was unable to figure this out.

Thanks and Regards.

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @ai21mtech13002 

Similar questions are answered here. Please take a look and let me know if you need more help. In that case, tell me what KD method you want to use.
  
https://github.com/yoshitomo-matsubara/torchdistill/issues/123
https://github.com/yoshitomo-matsubara/torchdistill/discussions/216
https://github.com/yoshitomo-matsubara/torchdistill/issues/222

**Reply by ai21mtech13002** [reply]
Thanks Yoshitomo,

I will get back to you if I have any problem. 

Thanks.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

Could you provide the classes/functions you added as `fasterrcnn_R18` and `fasterrcnn_R50`?

The screenshot of the yaml shows that you registered `fasterrcnn_R18` and `fasterrcnn_R50` by adding `@register_model_func` prior to the classes/functions, which do not accept the keyword argument `pretrained`.

If you made any other changes for your own models, please share the modified code (as text or file, not screenshot) here.

P.S. Reply in this thread to put everything per answer in one thread.
Your latest comment (with screenshots) is a follow-up question but seen as a "answer" candidate by GitHub.

**Reply by ai21mtech13002** [reply]
Hi,

1. I have followed the following steps to implement the Knowledge Distillation experiment:-
     (a). Created a Custom folder and added __init__.py, fasterrcnn_R50.py(the Teacher Class) and fasterrcnn_R18.py (for the Student class) using @register_model_class as was guided in #216.
     (b). Changed the YAML config files to include the VHR Dataset (with type as 'cocodetect' as I am using coco annotations) and the Teacher and Student models.
      (c). I have also included the pretrained model weight file for the Teacher (trained on VHR 10 dataset) in ckpt in the yaml file. 

2.  I am including the following files:
     (a) fasterrcnn_R50.py
[fasterrcnn_R50.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10050512/fasterrcnn_R50.txt)
      (b) fasterrcnn_R18.py
[fasterrcnn_R18.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10050515/fasterrcnn_R18.txt)
       (c) yaml config file
[config.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10050519/config.txt)

Thanks

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

Thank you for sharing the files.

I believe that you don't need either `fasterrcnn_R50` class or `fasterrcnn_resnet50_fpn.py`, but just use `fasterrcnn_resnet50_fpn` in yaml since
1. `fasterrcnn_R50` is identical to `fasterrcnn_resnet50_fpn` according to the following implementation and
2. `__init__` in either your `fasterrcnn_resnet50_fpn` or `fasterrcnn_R18` does not accept any keyword arguments except `self`, which is why you see "unexpected keyword argument" errors

```python
class fasterrcnn_R50(nn.Module):

    def __init__(self):
        self.model=torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=False)

    def forward(self,x):
        out=self.model(x)
        return out
```

e.g., According to your yaml config, the following parameters in `params` will be passed as keyword arguments to your teacher model, but the above implementation does not accept any of them.
```yaml
    params:
      pretrained: True
      progress: True
      num_classes: 11
      pretrained_backbone: True
```

Also, it looks like `fasterrcnn_R18` has some bug (conv1) and seems unnecessary complex (no need to register resnet modules to `self`), and I simplified it as a function to return Faster R-CNN with ResNet-18

```python
from torchdistill.models.registry import register_model_func
import torch.nn as nn
import torchvision
from torchvision.models.detection import FasterRCNN
from torchvision.models.detection.rpn import AnchorGenerator


@register_model_func
def fasterrcnn_resnet18_fpn():
    r18 = torchvision.models.resnet18(pretrained=True)
    backbone = nn.Sequential(r18.conv1, r18.bn1, r18.resnet18_relu, r18.resnet18_max_pool, r18.layer1, r18.layer2, r18.layer3, r18.layer4)

    anchor_generator = AnchorGenerator(sizes=((32, 64, 128, 256, 512),),aspect_ratios=((0.5, 1.0, 2.0),))
    # Feature maps to perform RoI cropping.
    # If backbone returns a Tensor, `featmap_names` is expected to
    # be [0]. We can choose which feature maps to use.
    roi_pooler = torchvision.ops.MultiScaleRoIAlign(featmap_names=['0'],output_size=7,sampling_ratio=2)
    # Final Faster RCNN model.
    model = FasterRCNN(backbone=backbone, num_classes=11, rpn_anchor_generator=anchor_generator, box_roi_pool=roi_pooler)
    return model
```

**Reply by ai21mtech13002** [reply]
Thanks for the detailed solution. I have been able to overcome the param error but am getting a different error this time. 
<img width="1422" alt="Screenshot 2022-11-21 at 9 14 40 PM" src="https://user-images.githubusercontent.com/104313548/203097444-ea20b3cc-c239-46a2-9123-9f52dff3cf5a.png">

This error happens after the stage I of Factor Transfer is done wherein only the teacher is trained and the stage II starts. Also enclosing the yaml config file used when the error occurred.

[config.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10057966/config.txt)


Thanks

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

Please keep the conversation in one thread (here).
Since the dataset is new to me, I would debug the dataset implementation, checking the following items:
1. [`targets`](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/object_detection.py#L127)
2. [`res`](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/object_detection.py#L135)

**Reply by ai21mtech13002** [reply]
HI, I have figured out the problem in the dataset, it was with the segmentation values in the annoatations. Using -test-only is giving me the detection results now. I have one query, in the 'ckpt' field of the student in the config ,do we give the path to a checkpoint for an EARLIER TRAINED student model or leave it blank so that the student model trains from scratch.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

The `ckpt` field is used to load and save the checkpoint of your model. If you give some file path that does not exist, the student model will be initialized without checkpoint, but the checkpoint will be saved there

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

Thanks for the clarification. I also have some more queries. 

1. In factor transfer the first stage (Stage I), is the training of the paraphraser governed by the reconstruction loss. So in this case what does the coco mAP metrics displayed at the end of first stage training represent? 

<img width="1428" alt="Screenshot 2022-11-23 at 12 15 34 PM" src="https://user-images.githubusercontent.com/104313548/203486655-98f08280-1ad4-4394-8d81-e2cf6cf16f2a.png">

3. Since the paraphraser is trained using the reconstruction loss, where can we see this value in the logs?

4. Also the sample config file specifies only 1 epoch for stage I. Wouldn't 1 epoch be too less for training the paraphraser?

Thanks.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

1. The COCO mAP values right after the first stage may be not meaningful as the student model is not trained at the moment, but only some auxiliary modules for the teacher model is trained.
2. The loss value is shown at the 2nd line in the screenshot.
3. See [the supplementary material of the FT paper](https://proceedings.neurips.cc/paper/2018/hash/6d9cb7de5e8ac30bd5e8734bc96a35c1-Abstract.html). They used 1 epoch for the 1st stage of training in their experiments with ImageNet (no hyperparameter description found for object detection, unfortunately)

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

1. Thanks for all your assistance. I was able to complete this experiment successfully. Sharing the final results:-
<img width="1418" alt="Screenshot 2022-11-23 at 6 53 53 PM" src="https://user-images.githubusercontent.com/104313548/203558618-8b000d3d-2ced-47f7-9fff-89f8b9b7513b.png">

2. I changed the optimiser for the paraphraser to Adam and trained it for 100 epochs. I then trained the student model in Stage II for 60 epochs.

3. I have some queries for you:-
(a) Is there some manner in which I can find the inference speed of the models trained using TorchDistill?
(b) Are there sample config files to run KD , Fitnets, Attention Transfer,KD via LSR and Knowledge Review especially for Object Detection. These would be really helpful to run more experiments.

4. Thanks a lot for all the assistance. This tool was really helpful. 

PS: Just a suggestion. If you could put comments in the config file, it would be really helpful for others using TorchDistill for the first time.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

Glad to hear it worked!

> (a) Is there some manner in which I can find the inference speed of the models trained using TorchDistill?

torchdistill does not offer a rigid inference time measurement. You can extract trained model weights from your checkpoint by

```shell
python3 examples/params_extractor.py --src <your_ckpt_file_path> --keys model --dst <output_model_weights_file_path>
```

> (b) Are there sample config files to run KD , Fitnets, Attention Transfer,KD via LSR and Knowledge Review especially for Object Detection. These would be really helpful to run more experiments.

For object detection tasks, some of them (e.g., Attention Transfer) may work simply by choosing intermediate layers from both student and teacher models to match like [this](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/single_stage/at/resnet18_from_resnet34.yaml#L106-L128), but I'm sure that either KD or FitNets would not work as logits matching between the student and teacher models is not doable.

Imagine M bounding boxes and N bounding boxes produced by student and teacher models respectively. The number of logits vectors would be equal to the number of bounding boxes.

Even if M = N, it is very unclear how we can find which bounding box from student should match one of the bounding boxes from teacher when computing KL Divergence.

I saw some papers claimed use of KD for object detection and asked them how they made it. But it was not KD actually e.g., See [this thread in the Knowledge Review repo](https://github.com/dvlab-research/ReviewKD/issues/10#issuecomment-890692357)

If you want to further discuss this topic, feel free to open a new discussion, and we can discuss it there.

Close this discussion by selecting [my initial reply](https://github.com/yoshitomo-matsubara/torchdistill/discussions/254#discussioncomment-4178940) as "answer" so that the community can tell that it's answered.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

I understand how KD and Fitnets would not work for the task of object detection. 

I gather that AT and Knowledge Review could be used and the method would be similar to the links as provided by you.(Am I right??)

Thanks.

**Reply by yoshitomo-matsubara** [reply]
Yes, that's correct

**Reply by ai21mtech13002** [reply]
Thanks

**Comment by ai21mtech13002** [comment]
Hi Yoshitomo,

I an trying to implement knowledge distillation on an object detection dataset(VHR 10 with 11 classes including the background
ground class) . I have trained the teacher (FasterRCNN with Resnet50 backbone) . I have implemented the steps as mentioned by you in #123 #216. However, i am running into the following error:-

<img width="1168" alt="Screenshot 2022-11-20 at 4 20 54 PM" src="https://user-images.githubusercontent.com/104313548/202898191-b039a58f-5351-4890-9444-da63c664e0ac.png">

I believe there must be some error in the declaration of the config file. I am attaching the screenshot below 

<img width="1070" alt="Screenshot 2022-11-20 at 4 23 29 PM" src="https://user-images.githubusercontent.com/104313548/202898326-da776877-6a29-46d1-915e-41e8c1e9342a.png">

Thanks

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

Is there any reason you cannot use the Faster R-CNN with ResNet-18 + FPN as a student model? 

Because `backbone` in your student model (using `nn.Sequential`) is different from Faster R-CNN with ResNet-18 + FPN in [this sample config](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/coco2017/multi_stage/ft/custom_fasterrcnn_resnet18_fpn_from_fasterrcnn_resnet50_fpn.yaml), you need to change paths of modules to be frozen. 

Also, it looks like your student model misses FPN, which is included in your teacher model (from torchvision), so it may not work even after you resolve the issue of frozen module paths.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

Yes I believe that was an error that I had not included FPN. 

I included the FPN in the student thereafter. I was able to run the experiment but it seems that the student model is not learning because the detection metrics(mAP) is 0 for all the epochs.

<img width="1389" alt="Screenshot 2022-11-22 at 9 05 57 PM" src="https://user-images.githubusercontent.com/104313548/203355960-4adf5f90-46c0-41dd-8b5c-a824e9bfad75.png">

I am including the log of the entire run.
[log.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10068838/log.txt)

I am also including the student(Resnet18+FPN) now being used       
[fasterrcnn_R18_FPN.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10068859/fasterrcnn_R18_FPN.txt)

and the .yaml config file (I have commented out parameters of the student that were not being passed in the custom resnet18+FPN model being used)
[config.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10068875/config.txt)

The pretrained teacher model is the same.

I would request you to please guide me where I seem to be going wrong.

Thanks

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

I believe that a big issue here is that your teacher model did not work properly. By default, you will see the teacher performance before you test a trained student model at the end of the training session. Here is the result in the log file.

```
2022/11/22 20:56:41	INFO	__main__	[Teacher: fasterrcnn_resnet50_fpn]
2022/11/22 20:56:42	INFO	torchdistill.misc.log	Test:  [  0/200]  eta: 0:03:16  model_time: 0.3919 (0.3919)  evaluator_time: 0.0071 (0.0071)  time: 0.9822  data: 0.5809  max mem: 19999
2022/11/22 20:57:07	INFO	torchdistill.misc.log	Test: Total time: 0:00:26
2022/11/22 20:57:07	INFO	__main__	Averaged stats: model_time: 0.0435 (0.1237)  evaluator_time: 0.0020 (0.0032)
2022/11/22 20:57:07	INFO	__main__	Accumulating evaluation results...
2022/11/22 20:57:07	INFO	__main__	DONE (t=0.09s).
2022/11/22 20:57:07	INFO	__main__	IoU metric: bbox
2022/11/22 20:57:07	INFO	__main__	 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
2022/11/22 20:57:07	INFO	__main__	 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.001
2022/11/22 20:57:07	INFO	__main__	 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
2022/11/22 20:57:07	INFO	__main__	 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
2022/11/22 20:57:07	INFO	__main__	 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
2022/11/22 20:57:07	INFO	__main__	 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
2022/11/22 20:57:07	INFO	__main__	 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.003
2022/11/22 20:57:07	INFO	__main__	 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.003
2022/11/22 20:57:07	INFO	__main__	 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.003
2022/11/22 20:57:07	INFO	__main__	 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
2022/11/22 20:57:07	INFO	__main__	 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
2022/11/22 20:57:07	INFO	__main__	 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.003
```

I'd first make sure that the teacher model achieves an expected performance (you can skip training and test the models by adding `-test_only` when running the script). If you've already confirmed that the teacher model weights are loaded properly, the dataset might be a root of the issue.

**Reply by ai21mtech13002** [reply]
I think the problem could be with the dataset. The VHR-10 dataset does not have any segmentation information. Therefore, after converting the annotations to the coco format, the segmentation values in the coco json file were "[]". Since, this was giving an error while running torchdistill, I have manually added segmentaion mask to each detection as specified in https://www.section.io/engineering-education/understanding-coco-dataset/#segmentation i.e.
<img width="712" alt="Screenshot 2022-11-22 at 11 53 52 PM" src="https://user-images.githubusercontent.com/104313548/203392962-360c9680-e3ca-42ab-a737-3d461f6dbb95.png">



If this is resulting in the error, how should i prepare the annotations?
